File: web/src/server/ingest.ts
import { readdir } from "node:fs/promises";
import type { ContribRecord } from "./types";
import { PUBLIC_MEMORIES_DIR } from "./paths";

export interface ExpectedAssets {
  splat: string;
  preview: string;
  thumbnail: string;
}

/** Filenames S1 + `npm run convert-splats` produce for a record id (by stem). */
export function expectedAssets(id: string): ExpectedAssets {
  return { splat: `${id}.sog`, preview: `${id}.preview.ply`, thumbnail: `${id}.jpg` };
}

export type IngestResult =
  | { ok: true; patch: Partial<ContribRecord> }
  | { ok: false; reason: string };

/**
 * Pure: given the set of filenames present in public/memories, decide whether a
 * record can transition to `ready` and which urls to attach. The splat is
 * required; the thumbnail is optional (used only for far billboards/UI).
 */
export function resolveIngest(id: string, present: ReadonlySet<string>): IngestResult {
  const assets = expectedAssets(id);
  if (!present.has(assets.splat)) {
    return { ok: false, reason: `splat ${assets.splat} not found in public/memories` };
  }
  return {
    ok: true,
    patch: {
      status: "ready",
      splat_url: assets.splat,
      thumbnail_url: present.has(assets.thumbnail) ? assets.thumbnail : "",
    },
  };
}

/** fs seam: list public/memories and run resolveIngest against it. */
export async function ingestFromDisk(id: string): Promise<IngestResult> {
  const present = new Set(await readdir(PUBLIC_MEMORIES_DIR));
  return resolveIngest(id, present);
}

File: backend/ingest.py
"""Decide whether a reconstructed record can go ``ready`` and which urls to attach.

Port of ``web/src/server/ingest.ts``. The splat is required; the thumbnail is optional
(used only for far billboards/UI).
"""
from __future__ import annotations

from typing import AbstractSet


def expected_assets(record_id: str) -> dict:
    """Filenames SHARP + convert-splats produce for a record id (by stem)."""
    return {
        "splat": f"{record_id}.sog",
        "preview": f"{record_id}.preview.ply",
        "thumbnail": f"{record_id}.jpg",
    }


def resolve_ingest(record_id: str, present: AbstractSet[str]) -> dict:
    """Given the filenames present in public/memories, decide the transition.

    Returns ``{"ok": True, "patch": {...}}`` or ``{"ok": False, "reason": str}``.
    """
    assets = expected_assets(record_id)
    if assets["splat"] not in present:
        return {"ok": False, "reason": f"splat {assets['splat']} not found in public/memories"}
    return {
        "ok": True,
        "patch": {
            "status": "ready",
            "splat_url": assets["splat"],
            "thumbnail_url": assets["thumbnail"] if assets["thumbnail"] in present else "",
        },
    }
